src/programs/city/defense.js

```diff
@@ -12,10 +12,73 @@ class CityDefense extends kernel.process {
       return this.suicide()
     }
 
-    this.safeMode()
+    const room = Game.rooms[this.data.room]
+
+    let towers = sos.lib.cache.getOrUpdate(
+      [this.data.room, 'towers'],
+      { persist: true, maxttl: 5000, chance: 0.01 },
+      () => room.find(FIND_MY_STRUCTURES, s => s.structureType === STRUCTURE_TOWER))
+
+    const hostiles = room.find(FIND_HOSTILE_CREEPS)
+
+    if (towers && towers.length > 0) {
+      this.fireTowers(towers, hostiles)
+    }
+
+    const playerHostiles = hostiles.filter(c => c.owner.username !== 'Invader')
+    if (playerHostiles.length > 0) {
+      this.safeMode(playerHostiles)
+    }
   }
 
-  safeMode () {
+  fireTowers (towers, hostiles) {
+    if (hostiles.length > 0) {
+      // for now, just shoot closest for each tower
+      for (let tower of towers) {
+        if (tower.energy < TOWER_ENERGY_COST) {
+          continue
+        }
+        const closest = _.min(hostiles, c => c.pos.getRangeTo(tower.pos))
+        tower.attack(closest)
+      }
+      return
+    }
+
+    const healFunc = (healTarget) => {
+      // TODO: clever calculations to avoid overheals
+      for (let tower of towers) {
+        tower.heal(healTarget)
+      }
+    }
+
+    if (this.data.healTarget !== undefined) {
+      const healTarget = Game.getObjectById(this.data.healTarget)
+      if (healTarget &&
+          (healTarget.pos.room.name === this.data.room) &&
+          (healTarget.hits < healTarget.hitsMax)) {
+        healFunc(healTarget)
+        return
+      }
+
+      // heal target no longer valid
+      delete this.data.healTarget
+    }
+
+    // look for a heal target every healFrequency ticks
+    const healFrequency = 5
+    if (Game.time % healFrequency === 0) {
```

Comment:
This is pretty dangerous on an OS model- if the defense program doesn't run that particular tick it'll be at least five more ticks before it hits (and it could possibly take longer). You may want to record the last tick this was run and then run this branch whenever `Game.time - this.data.lastHeal >= 5`.

Reply:
Usually I do that, but on the `defense` program I figure it's going to run every tick.

I'll add a helper for that, though.